go/new_merkletree/merkle_tree.go

```diff
@@ -0,0 +1,173 @@
+// Package new_merkletree provides a means to generate RFC6962-compliant
+// Merkle Hash Trees.
+package new_merkletree
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"hash"
+	"math"
+)
+
+// New creates a new merkle hash tree.  The number of "leaves" of the tree,
+// as well as their contents, are retrieved through |dao|.  If you wish to
+// have acceptable performance on non-trivial tree sizes, you'll want to
+// provide |cache| (otherwise, pass `nil`).  The hash function used for all
+// nodes in the tree is specified by |hasher|.
+func New(dao NewMerkleTreeDataInterface, cache NewMerkleTreeCacheInterface, hasher hash.Hash) *NewMerkleTree {
+	return &NewMerkleTree{dao: dao, cache: cache, hasher: hasher}
+}
+
+func (mt NewMerkleTree) CurrentRoot() (Hash, error) {
```

Comment:
Updated.  I'm still getting my head around exactly when pointer receivers should be used, so thanks for the, uh, pointer.  (grin)